instrumentation/jdbc/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/jdbc/JdbcAdviceScope.java

```diff
@@ -0,0 +1,114 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.javaagent.instrumentation.jdbc;
+
+import static io.opentelemetry.javaagent.bootstrap.Java8BytecodeBridge.currentContext;
+import static io.opentelemetry.javaagent.instrumentation.jdbc.JdbcSingletons.statementInstrumenter;
+
+import io.opentelemetry.context.Context;
+import io.opentelemetry.context.Scope;
+import io.opentelemetry.instrumentation.jdbc.internal.DbRequest;
+import io.opentelemetry.instrumentation.jdbc.internal.JdbcData;
+import io.opentelemetry.javaagent.bootstrap.CallDepth;
+import java.sql.PreparedStatement;
+import java.sql.Statement;
+import java.util.Map;
+import javax.annotation.Nullable;
+
+public class JdbcAdviceScope {
+  private final CallDepth callDepth;
+  private final DbRequest request;
+  private final Context context;
+  private final Scope scope;
+
+  private JdbcAdviceScope(CallDepth callDepth, DbRequest request, Context context, Scope scope) {
+    this.callDepth = callDepth;
+    this.request = request;
+    this.context = context;
+    this.scope = scope;
+  }
+
+  public static JdbcAdviceScope startBatch(CallDepth callDepth, Statement statement) {
+    return start(callDepth, null, statement, null);
+  }
+
+  public static JdbcAdviceScope startStatement(
+      CallDepth callDepth, String sql, Statement statement) {
+    return start(callDepth, sql, statement, null);
+  }
+
+  public static JdbcAdviceScope startPreparedStatement(
+      CallDepth callDepth, PreparedStatement preparedStatement) {
+    return start(callDepth, null, null, preparedStatement);
+  }
+
+  private static JdbcAdviceScope start(
+      CallDepth callDepth,
+      @Nullable String sql,
+      @Nullable Statement statement,
+      @Nullable PreparedStatement preparedStatement) {
+    // Connection#getMetaData() may execute a Statement or PreparedStatement to retrieve DB info
+    // this happens before the DB CLIENT span is started (and put in the current context), so this
+    // instrumentation runs again and the shouldStartSpan() check always returns true - and so on
+    // until we get a StackOverflowError
+    // using CallDepth prevents this, because this check happens before Connection#getMetadata()
+    // is called - the first recursive Statement call is just skipped and we do not create a span
+    // for it
+    if (callDepth.getAndIncrement() > 0) {
+      return new JdbcAdviceScope(callDepth, null, null, null);
+    }
+
+    Context parentContext = currentContext();
+
+    DbRequest request;
+    if (sql != null) {
+      request = DbRequest.create(statement, sql);
+    } else if (preparedStatement != null) {
+      Map<String, String> parameters = JdbcData.getParameters(preparedStatement);
+      request = DbRequest.create(preparedStatement, parameters);
+    } else {
+      // batch
+      request = createBatchRequest(statement);
+    }
```

Comment:
perhaps it would be better for `start` to accept `Supplier<DbRequest>` so that the request creation could be moved to `startStatement` etc.

Reply:
I was thinking of pushing the `Supplier` creation directly into the advice, but you are right keeping it `JdbcAdviceScope` already makes things easier to read. Once everything is running with indy advices we could even directly remove those 3 intermediate methods that are called from advices.